[0052] Extensible UDT (#428)
nervosnetwork/rfcs: rfcs/0052-extensible-udt/0052-extensible-udt.md

```

`is_owner_mode` indicates if the current xUDT is unlocked via owner mode(as
described by sUDT), `extension_index` refers to the index of the current
extension in the `ScriptVec` structure. `args` and `args_length` are set to the
script args included in `Script` structure of the current extension script.

If this function returns 0, the current extension script validation is
considered successful.

• If `flags & 0x1FFFFFFF` is 0x2, extension data will contain the blake160 hash
of the `ScriptVec` structure as explained in the previous section. The
actual `raw_extension_data` (`ScriptVec`) structure data will be included in a
witness field `input_type` or `output_type` contained in the current
transaction. We will explain this part below. Choosing `input_type` or
`output_type` depends on whether the type script is running on input or output
cells. Under a lot of scenarios, it is `input_type`. But in the following example “Owner
Mode Without Consuming Cell”, we can see it’s possible on `output_type`.

### xUDT Witness

The `input_type` or `output_type` field in witness has the following data
structure in molecule format:

```js

> table XudtWitnessInput {

Reviewer: The name here is slightly confusing. While most cases would keep this data in `input_type` field, there are cases where it might be stored in `output_type` field. This means we probably should not name it `XXXInput`.

Author: Updating this name is easy. But the script code is freezed and the script is deployed.
Can we just add some comments about it's possible in `output_type` in some scenarios.

Reviewer: I do understand the code is freezed, I'm not suggesting any change in behaviors.

But I do believe that API-side, we can tweak the name without any breaking code. There is even a chance the binary code would still stay the same even if we modify the field names in molecule source file, recreate molecule generated file, then recompile the C source file, assuming we've done sufficient purging. Since the RFC is still being written, I do recommend we spend enough time thinking about proper names.

    owner_script: ScriptOpt,		
    owner_signature: BytesOpt,
    raw_extension_data: ScriptVecOpt,
    extension_data: BytesVec,
}
```

The field `owner_script` and `owner_signature` will be used in owner mode. The

> field `raw_extension_data` and `extension_data` are used when `flags &

Reviewer: As explained below, `extension_data` is not only used here, it is used in previous cases as well.

Reviewer: And the naming choice here is also misleading. One would think that `raw_extension_data` and `extension_data` are correlated, but in fact they are used in different places.

0x1FFFFFFF` is 0x2 in args.

### Owner Mode Update

As described in RFC sUDT, if an input cell in the current transaction uses an
input lock script with the same script hash as the owner lock script hash, the
`is_owner_mode` will be set to true. In xUDT, this rule is updated by the
following rule:

If an input or output cell in the current transaction uses one or more of the
following:

- input lock script (when `flags & 0x20000000` is **zero** or `flags` is not present)
- output type script (when `flags & 0x40000000` is **non-zero**)
- input type script (when `flags & 0x80000000` is **non-zero**)

With the same script hash as the owner lock script hash, the `is_owner_mode`
will be set to true. The output lock scripts are not included, because they
won’t be run in a transaction.

If the `owner_script` in witness isn’t none and its blake2b hash is the same as
the owner lock script hash in `args`, this script will be run as an extension
script. If the script returns success, `is_owner_mode` is set to true. Note, the

> `owner_signature` field can be used by this owner script. Usually, a valid

Reviewer: Calling this field `xxx_signature`, when the text explicitly mentions this field **usually** contains a signature(hinting there might be case this field does not contain a signature) is so weird.

all governance operations, such as minting.

A normal transfer operation of xUDT, however, differs from Simple UDT. Depending
on the flags used, there might be 2 usage patterns:

### **Raw Extension Script**

When `flags & 0x1FFFFFFF` are set to 0x1, raw extension data is included in xUDT args directly.

```yaml
Inputs:
    <vec> xUDT_Cell
        Data:
            <amount: uint128> <xUDT data>
        Type:
            code_hash: xUDT type script
            args: <owner lock script hash> <xUDT args>
        Lock:
            <user defined>
    <...>
Outputs:
    <vec> xUDT_Cell
        Data:
            <amount: uint128> <xUDT data>
        Type:
            code_hash: xUDT type script
            args: <owner lock script hash> <xUDT args>
        Lock:
            <user defined>
    <...>
Witnesses:
    WitnessArgs structure:
        Lock: <user defined>
        Input Type: <XudtWitnessInput>
            owner_script: <None>
            owner_signature: <None>				
            raw_extension_data: <None>
            extension_data: 
                <vec> BytesVec
                    <data> 
                <...>
```

> The witness of the same index as the first input xUDT cell is located by xUDT script. It is parsed first as WitnessArgs structure, the `input_type` or `output_type` field of `WitnessArgs`, is thus treated as `XudtWitnessInput` structure. The length of `extension_data` structure inside must also be the same as `ScriptVec` in `xUDT args`. An extension script might also require transaction-specific data for validation. The witness here provides a place for these data needs.

Reviewer: I recommend we have a dedicated place explaining the purpose of `extension_data`, instead of hiding it deeply here.

> Note that each extension script is only executed once in the transaction. The extension script is responsible for checking all xUDT cells of the current type, ensuring each cell data and witness for the current extension script, can be validated against the extension script’s rules.

Reviewer: > Note that each extension script is only executed once in the transaction.

Is this ensured in the implementation? What if the same extension script is included multiple times on purpose? If so, how?

Author: > each extension script is only executed once
It means every extension script is executed one by one, only once. Same extension scripts included multiple times will run for multiple times. There is no optimization.

Author: When multiple instances of the same extension script are included, each instance will execute independently for each inclusion.

### **P2SH Style Extension Script**

When `flags & 0x1FFFFFFF` are set to 0x2, only the blake160 hash of extension data is included in xUDT args. The user is required to provide the actual extension data in witness directly:

```yaml
Inputs:
    <vec> xUDT_Cell
        Data:
            <amount: uint128> <xUDT data>
        Type:
            code_hash: xUDT type script

>             args: <owner lock script hash> <xUDT args>

Reviewer: Better to mark clearly that the `xUDT args` field here contains a hash, not the actual value.

hash as the owner lock script hash, the `is_owner_mode` will be set to true. It
isn’t convenient: this requires extra cell to be consumed. With `owner_script`
and `owner_signature` set to proper values, we can use owner mode without extra
cell.

```yaml
Inputs:
    <vec>
			<Any input cells>
    <...>
Outputs:
    <vec> xUDT_Cell
        Data:
            <amount: uint128> <xUDT data>
        Type:
            code_hash: xUDT type script
            args: <owner lock script hash 1> <xUDT args>
        Lock:
            <user defined>
    <...>
Witnesses:
    WitnessArgs structure:
        Lock: <user defined>
        Input Type: <None>
        Output Type: XudtWitnessInput
            owner_script: <owner script 1>
            owner_signature: <signature 1>				
            raw_extension_data: 
                <vec> ScriptVec
                    <script>
                <...>
            extension_data: 
                <vec> BytesVec
                    <data>
                <...>
```

The example above shows a scenario of owner mode without consuming the owner's cell. The `<owner lock script
hash 1>` is the same as blake160 hash of `<owner script 1>`. We can implement an

> extension script as `<owner script 1>` which has same functionality to

Reviewer: Can we rephrase this part? It's so misleading. How can we implement such a thing? Or is it required to implement such a script?

---
Number: "0052"
Category: Standards Track
Status: Proposal
Author: Xuejie Xiao <xxuejie@gmail.com>, Xu Jiandong <lynndon@gmail.com>
Created: 2024-01-09
---

# Extensible UDT

Extensible UDT(xUDT) is an extension of [Simple
UDT](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0025-simple-udt/0025-simple-udt.md) for
defining more behaviors a UDT might need. While simple UDT provides a minimal
core for issuing UDTs on Nervos CKB, extensible UDT builds on top of simple UDT
for more potential needs, such as regulations.

## **Data Structure**

**xUDT Cell**

An xUDT cell is backward compatible with Simple UDT, all the existing rules
defined in the Simple UDT spec must still hold true for xUDT cells. On top of
sUDT, xUDT extends a cell like the following:

```yaml
data:
    <amount: uint128> <xUDT data>
type:

>     code_hash: xUDT type script
>     args: <owner lock script hash> <xUDT args>

Reviewer 2: ```suggestion
    code_hash: xUDT type script
    hash_type: data1 on mainnet | type on testnet
    args: <owner lock script hash> <xUDT args>
```

In the current deployments, developers using xUDT will have to switch the `hash_type` of xUDT type_script between testnet and mainnet.

Author: In spec, we can't write that. We can deploy this script on local devnet with other code_hash/hash_type.

lock:
    <user_defined>
```

The added `xUDT args` and `xUDT data` parts provide all the new functions needed

> by xUDT, we will explain the detailed structure below.

Reviewer 2: ```suggestion
by xUDT, the detailed structure is explained below.
```

### **xUDT Args**

xUDT args has the following structure:

```
<4-byte xUDT flags> <Variable length bytes, extension data>
```

Depending on the content of `flags`, different extension data might be attached:

> • If `flags & 0x1FFFFFFF` is 0, we won’t need any extension data. Note a

Reviewer 2: ```suggestion
• If `flags & 0x1FFFFFFF` is 0, no extension data is required. Note a
```

backward-compatible way of viewing things, which is that a plain sUDT cell also
has a hidden `flags` field with all zeros.

• If `flags & 0x1FFFFFFF` is 0x1, extension data will contain a
[molecule](https://github.com/nervosnetwork/molecule) serialized `ScriptVec`
structure:

```
table Script {
    code_hash:      Byte32,
    hash_type:      byte,
    args:           Bytes,
}

vector ScriptVec <Script>
```

> Each entry included in `ScriptVec` structure is interpreted as a CKB script hash
> for an extension script with additional behaviors. When an xUDT script is

Reviewer 2: ```suggestion
Each entry included in `ScriptVec` structure is interpreted as
an extension script with additional behaviors. When an xUDT script is
```

If this function returns 0, the current extension script validation is
considered successful.

• If `flags & 0x1FFFFFFF` is 0x2, extension data will contain the blake160 hash
of the `ScriptVec` structure as explained in the previous section. The
actual `raw_extension_data` (`ScriptVec`) structure data will be included in a
witness field `input_type` or `output_type` contained in the current
transaction. We will explain this part below. Choosing `input_type` or
`output_type` depends on whether the type script is running on input or output
cells. Under a lot of scenarios, it is `input_type`. But in the example “Owner
Mode Without Consuming Cell”, we can see it’s possible on `output_type`.

### xUDT Witness

The `input_type` or `output_type` field in witness has the following data
structure in molecule format:

```js
table XudtWitnessInput {
    owner_script: ScriptOpt,		
    owner_signature: BytesOpt,
    raw_extension_data: ScriptVecOpt,
    extension_data: BytesVec,
}
```

The field `owner_script` and `owner_signature` will be used in owner mode. The
field `raw_extension_data` and `extension_data` are used when `flags &
0x1FFFFFFF` is 0x2 in args.

### Owner Mode Update

> As described in RFC sUDT, If an input cell in the current transaction uses an

Reviewer 2: ```suggestion
As described in RFC sUDT, if an input cell in the current transaction uses an
```

> As described in RFC sUDT, If an input cell in the current transaction uses an
> input lock script with the same script hash as the owner lock script hash, the
> `is_owner_mode` will be set to true. In xUDT, this rule is updated by the
> following rule:

Reviewer 2: ```suggestion
As described in RFC sUDT, if an input cell in the current transaction has
the same lock script hash with the owner's lock script hash, the
`is_owner_mode` will be set to true. In xUDT, this rule is updated by the
following rule:
```

Author: After updating, it reads: "an input cell has same lock script hash" which is incorrect.

Reviewer 2: yep, the original sentence is better.

is reserved for lock script specific data for current cells.

An extension script will first need to first locate the index it resides in xUDT
args, then look for the data for the current extension script at the same index
in `data` field of `XudtData` structure.

## **Operations**

xUDT uses the same governance operations as Simple UDT: an owner lock controls
all governance operations, such as minting.

A normal transfer operation of xUDT, however, differs from Simple UDT. Depending
on the flags used, there might be 2 usage patterns:

### **Raw Extension Script**

When `flags & 0x1FFFFFFF` are set to 0x1, raw extension data is included in xUDT args directly.

```yaml
Inputs:
    <vec> xUDT_Cell
        Data:
            <amount: uint128> <xUDT data>
        Type:
            code_hash: xUDT type script
            args: <owner lock script hash> <xUDT args>
        Lock:
            <user defined>
    <...>
Outputs:
    <vec> xUDT_Cell
        Data:
            <amount: uint128> <xUDT data>
        Type:
            code_hash: xUDT type script
            args: <owner lock script hash> <xUDT args>
        Lock:
            <user defined>
    <...>

> Witnesses:
>     WitnessArgs structure:
>     Lock: <user defined>
>     Input Type: <XudtWitnessInput>
>         owner_script: <None>
>         owner_signature: <None>				
>         raw_extension_data: <None>
>         extension_data: 
>             <vec> BytesVec
>                 <data> 
>             <...>

Reviewer 2: ```suggestion
Witnesses:
    WitnessArgs structure:
        Lock: <user defined>
        Input Type: <XudtWitnessInput>
            owner_script: <None>
            owner_signature: <None>				
            raw_extension_data: <None>
            extension_data: 
                <vec> BytesVec
                    <data> 
                <...>
```
add indents

```

> The witness of the same index as the first input xUDT cell is located by xUDT script. It is parsed first as WitnessArgs structure, the `input_type` or `output_type` field of `WitnessArgs`, is thus treated as `XudtWitnessInput` structure. The `extension_data`structure inside must also be of the same length as `xUDT args`. An extension script might also require transaction-specific data so as to validate. Witness here provides a place for this data needs.

Reviewer 2: ```suggestion
The witness of the same index as the first input xUDT cell is located by xUDT script. It is parsed first as WitnessArgs structure, the `input_type` or `output_type` field of `WitnessArgs`, is thus treated as `XudtWitnessInput` structure. The length of `extension_data` structure inside must also be the same as `ScriptVec` in `xUDT args`. An extension script might also require transaction-specific data for validation. The witness here provides a place for these data needs.
```

> Notice each extension script is only executed once in the transaction. The extension script is responsible for checking all xUDT cells of the current type, ensuring each cell data and witness for the current extension script, can be validated per the extension script’s rules.

Reviewer 2: ```suggestion
Note that each extension script is only executed once in the transaction. The extension script is responsible for checking all xUDT cells of the current type, ensuring each cell data and witness for the current extension script, can be validated against the extension script’s rules.
```

            code_hash: xUDT type script
            args: <owner lock script hash> <xUDT args>
        Lock:
            <user defined>
    <...>
Outputs:
    <vec> xUDT_Cell
        Data:
            <amount: uint128> <xUDT data>
        Type:
            code_hash: xUDT type script
            args: <owner lock script hash> <xUDT args>
        Lock:
            <user defined>
    <...>
Witnesses:
    WitnessArgs structure:
        Lock: <user defined>
        Input Type: XudtWitnessInput
            owner_script: <None>
            owner_signature: <None>				
            raw_extension_data: 
                <vec> ScriptVec
                    <script>
                <...>
            extension_data: 
                <vec> BytesVec
                    <data>
                <...>
```

The only difference here is that `XudtWitnessInput` in `input_type` or
`output_type` field in the corresponding WitnessArgs structure, contains raw
extension data in `ScriptVec` data structure, xUDT script must first validate
that the hash of raw extension data provide here, is the same as blake160 hash
included in xUDT args. After this, it uses the same logic as the previous
workflow.

> ### Owner Mode Without Consuming Cell

Reviewer 2: ```suggestion
### Owner Mode without Consuming Cell
```

> As described above, If an input cell uses an input lock script with same script
> hash as the owner lock script hash, the `is_owner_mode` will be set to true. It

Reviewer 2: ```suggestion
As described above, If an input cell in the current transaction has the same
lock script hash as the owner's lock script hash, the `is_owner_mode` will be set to true. It
```

> isn’t convenience: this requires extra cell to be consumed. With `owner_script`

Reviewer 2: ```suggestion
isn’t convenient: this requires extra cell to be consumed. With `owner_script`
```

> and `owner_signature` with proper values, we can use owner mode without extra

Reviewer 2: ```suggestion
and `owner_signature` set to proper values, we can use owner mode without extra
```

                <...>
            extension_data: 
                <vec> BytesVec
                    <data>
                <...>
```

The example above shows a scenario of owner mode without consuming cell. This
transaction can be valid even without any input cell. The `<owner lock script
hash 1>` is the same as blake160 hash of `<owner script 1>`. We can implement an
extension script as `<owner script 1>` which has same functionality to
[secp256k1/blake160 lock
script](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0024-ckb-genesis-script-list/0024-ckb-genesis-script-list.md#secp256k1blake160).
The `<signature 1>` can be used by `<owner script 1>` to place signature
information.

## Deployment

An [implementation](https://github.com/nervosnetwork/ckb-production-scripts/blob/master/c/xudt_rce.c) of
the spec above has been deployed to Mirana CKB mainnet and Pudge testnet:

- Mirana(mainnet)

| parameter | value |
| --- | --- |
| code_hash | 0x50bd8d6680b8b9cf98b73f3c08faf8b2a21914311954118ad6609be6e78a1b95 |
| hash_type | data1 |
| tx_hash | 0xc07844ce21b38e4b071dd0e1ee3b0e27afd8d7532491327f39b786343f558ab7 |
| index | 0x0 |
| dep_type | code |

> This script is not upgradeable due to zero lock (lock args with all zeros).

Reviewer 2: When I see the word `zero lock`, I thought it is a lock whose `code_hash` is all zero, not just with all zeros lock args.

## [Some ckb-system-scripts](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0024-ckb-genesis-script-list/0024-ckb-genesis-script-list.md) such as SECP256K1 and Nervos DAO are using lock script
```json
{
  "code_hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "hash_type": "data",
  "args": "0x"
}
```

## While xUDT is using
```json
"a_lock_with_zero_args": {
  "code_hash": "0x9bd7e06f3ecf4be0f2fcd2188b23f1b9fcc88e5d4b65a8637b17723bbda3cce8",
  "hash_type": "type"
  "args": "0x0000000000000000000000000000000000000000",
},
```

I know it is almost impossible to unlock `a_lock_with_zero_args`.
Just got a little confused with the `zero lock` wording.

Author: It's limited by ckb-cli deployment tools. It can't set code_hash to all zero.

The `data` field included in `XudtData`, must be of the same length
as `ScriptVec` structure included in xUDT args. Some extensions might require
user-specific data stored in each xUDT cell. xUDT data provides a place for such

> data. This XUDT data part can be optional.

Reviewer 3: Nice!
